### PR TITLE
Feature fast decimal to double

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

**Comment by sopel39** [comment]
```
Before:

Benchmark                                                          (expression)   Mode  Cnt  Score   Error   Units
BenchmarkDecimalOperators.decimalToDoubleCastBenchmark  cast(l_38_30 as double)  thrpt   30  0,044 ± 0,002  ops/ms
BenchmarkDecimalOperators.decimalToDoubleCastBenchmark   cast(l_38_2 as double)  thrpt   30  0,433 ± 0,018  ops/ms
BenchmarkDecimalOperators.decimalToDoubleCastBenchmark   cast(s_17_9 as double)  thrpt   30  2,027 ± 0,086  ops/ms

After:

Benchmark                                                          (expression)   Mode  Cnt  Score   Error   Units
BenchmarkDecimalOperators.decimalToDoubleCastBenchmark  cast(l_38_30 as double)  thrpt   30  0,050 ± 0,002  ops/ms
BenchmarkDecimalOperators.decimalToDoubleCastBenchmark   cast(l_38_2 as double)  thrpt   30  1,127 ± 0,039  ops/ms
BenchmarkDecimalOperators.decimalToDoubleCastBenchmark   cast(s_17_9 as double)  thrpt   30  1,962 ± 0,077  ops/ms
```

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 1]
```diff
@@ -498,6 +516,41 @@ public static long hash(long rawLo, long rawHi)
...
+        boolean negative = isNegative(decimal);
+
+        // convert decimal to strings one digit group at a time
+        do {
+            int remainder = divide(decimal, divisor, decimal);
+            digitGroup[numGroups++] = Integer.toString(remainder, 10);
+        }
+        while (!isZero(decimal));
+
+        StringBuilder buf = new StringBuilder(MAX_PRECISION);
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: MAX_PRECISION + 1?

File: presto-spi/src/test/java/com/facebook/presto/spi/type/TestUnscaledDecimal128Arithmetic.java [hunk 2]
```diff
@@ -246,6 +247,19 @@ public void testHash()
...
+    {
+        assertEquals(toStringDestructive(unscaledDecimal(0)), "0");
+        assertEquals(toStringDestructive(negateConstructive(unscaledDecimal(0))), "0");
+        assertEquals(toStringDestructive(unscaledDecimal(1)), "1");
+        assertEquals(toStringDestructive(unscaledDecimal(-1)), "-1");
+        assertEquals(toStringDestructive(MAX_DECIMAL), MAX_DECIMAL_UNSCALED_VALUE.toString());
+        assertEquals(toStringDestructive(MIN_DECIMAL), MIN_DECIMAL_UNSCALED_VALUE.toString());
+        assertEquals(toStringDestructive(unscaledDecimal("1000000000000000000000000000000000000")), "1000000000000000000000000000000000000");
+        assertEquals(toStringDestructive(unscaledDecimal("-1000000000002000000000000300000000000")), "-1000000000002000000000000300000000000");
+    }
```
Comment (kokosing) [inline presto-spi/src/test/java/com/facebook/presto/spi/type/TestUnscaledDecimal128Arithmetic.java:261]: add test for decimal with fraction part

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 3]
```diff
@@ -498,6 +516,41 @@ public static long hash(long rawLo, long rawHi)
         return XxHash64.hash(rawLo) ^ XxHash64.hash(unpackUnsignedLong(rawHi));
     }
 
+    public static String toStringDestructive(Slice decimal)
+    {
+        if (isZero(decimal)) {
+            return "0";
+        }
+
+        int divisor = POWERS_OF_TEN_INT[MAX_POWER_OF_TEN_INT];
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: name it that is biggest power of ten

**Comment by kokosing** [comment]
looks good

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 1 above)
Comment (sopel39) [reply presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: done

File: presto-spi/src/test/java/com/facebook/presto/spi/type/TestUnscaledDecimal128Arithmetic.java (hunk 2 above)
Comment (sopel39) [reply presto-spi/src/test/java/com/facebook/presto/spi/type/TestUnscaledDecimal128Arithmetic.java:261]: this only outputs textual representation for unscaled value, no fraction

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 3 above)
Comment (sopel39) [reply presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: done

File: presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java [hunk 4]
```diff
@@ -404,10 +429,12 @@ public static double shortDecimalToDouble(long decimal, long precision, long sca
...
-        // TODO: optimize
-        BigInteger decimalBigInteger = decodeUnscaledValue(decimal);
-        BigDecimal bigDecimal = new BigDecimal(decimalBigInteger, (int) scale);
-        return bigDecimal.doubleValue();
+        // If both decimal and scale can be represented exactly in double then compute rescaled and rounded result directly in double.
+        if (scale < DOUBLE_10_POW.length && compareUnsigned(decimal, MAX_EXACT_DOUBLE) <= 0) {
+            return (double) unscaledDecimalToUnscaledLongUnsafe(decimal) / DOUBLE_10_POW[(int) scale];
+        }
+
+        return parseDouble(Decimals.toString(decimal, (int) scale));
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java:437]: add comment:

`TODO: optimize and convert directly to double in similar fashion as in double to decimal casts`

File: presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java [hunk 5]
```diff
@@ -419,10 +446,12 @@ public static long shortDecimalToFloat(long decimal, long precision, long scale,
...
-        // TODO: optimize
-        BigInteger decimalBigInteger = decodeUnscaledValue(decimal);
-        BigDecimal bigDecimal = new BigDecimal(decimalBigInteger, (int) scale);
-        return floatToRawIntBits(bigDecimal.floatValue());
+        // If both decimal and scale can be represented exactly in float then compute rescaled and rounded result directly in float.
+        if (scale < FLOAT_10_POW.length && compareUnsigned(decimal, MAX_EXACT_FLOAT) <= 0) {
+            return floatToRawIntBits((float) unscaledDecimalToUnscaledLongUnsafe(decimal) / FLOAT_10_POW[(int) scale]);
+        }
+
+        return floatToRawIntBits(parseFloat(Decimals.toString(decimal, (int) scale)));
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java:454]: add comment:

`TODO: optimize and convert directly to double in similar fashion as in double to decimal casts`

**Pull request closed without merging**